src/file-transfer-handler.js

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import UploadOptionsBase from './upload-options-base';
+import HttpClient from './http/http-client';
+import FileUploadResult from './file-upload-result';
+
+const DEFAULT_PROGRESS_THRESHOLD = 500;
+
+/**
+ * The purpose of this class is to coordinate the complete transfer of a file,
+ * based on a process that transfers said file in multiple parts.
+ *
+ * This class will handle determining when a file starts transferring, and
+ * when a file finishes transferring, either successfully or in failure.
+ */
+export default class FileTransferHandler extends UploadOptionsBase {
```

Comment:
I think the inheritance here may be justified if `ConcurrentQueue` is NOT a `sister` class and is moved in another hierarchy.
(I would still consider either renaming `UploadOptionsBase` to `UploaderBase`, or turn `UploadOptionsBase` into an attribute, so composed inside that class).

Reply:
> turn UploadOptionsBase into an attribute, so composed inside that class

Could you expound on this? Do you have an example of what you'd propose?

Reply:
What I would like to use it like as a library user:

```
const uploadOptions = new UploadOptionsBase(`put something here`);
const fileTransferHandler = new FileTransferHandler(uploadOptions);
```

And inside the classes, to get access to the functions from type `UploadOptionsBase` something like:
```
this.uploadOptions.log_this_for_me('log_me');
```

Reply:
I see. So rather than having the common functionality in a base class, have the common functionality in an object that given to the classes that need it. That's a great idea.

Like I mentioned above, I don't want to balloon this changeset more, so I'll make a note to do some refactoring on my next iteration. Thanks for the suggestion!